library/crypto.php

```diff
@@ -83,36 +87,34 @@ function cryptCheckStandard($value)
  *
  * @param  string  $sValue          Raw data that will be encrypted.
  * @param  string  $customPassword  If null, then use standard key. If provide a password, then will derive key from this.
+ * @param  string  $keySource       This is the source of the keys. Options are 'drive' and 'database'
  * @return string                   returns the encrypted data.
  */
-function aes256Encrypt($sValue, $customPassword = null)
+function aes256Encrypt($sValue, $customPassword = null, $keySource = 'drive')
 {
     if (!extension_loaded('openssl')) {
         error_log("OpenEMR Error : Encryption is not working because missing openssl extension.");
     }
 
     if (empty($customPassword)) {
-        // Collect the encryption keys. If they does not exist, then create them
+        // Collect the encryption keys. If they do not exist, then create them
         // The first key is for encryption. Then second key is for the HMAC hash
-        $sSecretKey = aes256PrepKey("two", "a");
-        $sSecretKeyHmac = aes256PrepKey("two", "b");
+        $sSecretKey = aes256PrepKey("four", "a", $keySource);
+        $sSecretKeyHmac = aes256PrepKey("four", "b", $keySource);
     } else {
-        // Turn the password into a hash(note use binary) to use as the keys
-        $sSecretKey = hash("sha256", $customPassword, true);
-        $sSecretKeyHmac = $sSecretKey;
+        // customPassword mode, so turn the password into hashes (note use binary) to use as the keys
+        // The first key is for encryption. Then second key is for the HMAC hash
+        $sSaltKey = produceRandomBytes(32);
+        $sSecretKey = hash_pbkdf2('sha384', $customPassword, $sSaltKey, 100000, 32, true);
+        $sSaltHmac = produceRandomBytes(32);
+        $sSecretKeyHmac = hash_pbkdf2('sha384', $customPassword, $sSaltHmac, 100000, 32, true);
```

Comment:
TODO
Waiting to see if creating 2 separate hashes/keys for passphrase mode is overkill.